### ISSUE TITLE
Tag datasets with machine-readable metadata

### ISSUE THREAD

**Issue by brawer** [alltheplaces/alltheplaces#4338]
Should spiders be able to declare dataset attributes (“metadata”), in particular for licensing?

ATP could export them as FeatureCollection properties. For data consumers, this would mean they could ingest ATP data while filtering out any datasets whose license they don’t like. Without machine-readable licensing data, consumers might be more reluctant to consume ATP data. Of course, most websites don’t actually declare any license; data consumers need to decide for themselves what they want to do in that case. Still, there is plenty of open data around with a clear license.

A scraper could optionally declare its dataset attributes like this:

```python
class ExampleSpider(scrapy.Spider):
    name = "example"
    allowed_domains = ["data.example.org"]
    dataset_attributes = {
        "license": "Creative Commons Zero",
        "license:website": "https://example.org/legal/",
        "license:wikidata": "Q6938433",
    }
```

The scraped GeoJSON file might look like this:

```json
{"type":"FeatureCollection","features":[
{"type": "Feature", "id": "RMOK0O_OUPw-GrmVyGJ_i6M266o=", "properties": {"name": "A", ...}},
{"type": "Feature", "id": "jy5en3B92fHIc70K8Gq2pGYCpKw=", "properties": {"name": "B", ...}}],
"properties":{"license":"Creative Commons Zero","license:website":"https://example.org/legal/","license:wikidata":"Q6938433"}}
```

I made an experimental implementation to see how this would fit into the ATP/Scrapy framework; pull request coming.

**Comment by mjoe999** [comment]
I was thinking of suggesting that we add license tags to the output but here you are with a proposal. Happy days.

So there's a few things going on here. I guess we have typically normalised the POI / dataset attributes within each "Feature". It's simple but bloated. I don't know how fussy GeoJSON clients are but I would think/hope not, so no problem to have such attributes common as you suggest. A small discussion I found here is of interest:

https://gis.stackexchange.com/questions/25279/is-it-valid-to-have-a-properties-element-in-an-geojson-featurecollection

I would suggest that the dataset attributes where written before the Feature elements? That way for a huge file you can stream process the JSON and get the meta-data before the items.

I guess the other thing I would suggest but not necessary at this stage is that we have a licence.py or similar which enumerates the common licences? That way in the spider we could have:

**licence = Licence.CREATIVE_COMMONS_ZERO**

Spiders would still be free to add their own licence terms but in most instances this enumeration would suffice. A bit like the categories.

@iandees might want to think about adding a default licence and / or some text / link-back to meta properties which points back to the project? i.e. a single GeoJSON output file is more self-describing?

Fundamentally this is a good thing you suggest.

**Comment by iandees** [comment]
This is a good idea, thanks for the suggestion!

All of the data we produce is currently CC-0 as mentioned on our readme: https://github.com/alltheplaces/alltheplaces#license, so we would need to find a way to communicate the fact that some data is now not CC-0. Some options: 

1. Existing download bundles stop updating
2. Existing download bundles only contain the CC-0 data. 

Then we could put together new "v2" download links that are filtered to "classes" of license like "cc-0", "share-alike", "other open".

Multiple license downloads are important for OpenAddresses support, by the way.

**Comment by brawer** [comment]
Curious, do data consumers actually need separate download bundles? As long as each feed is tagged with its license, I’d have assumed they can filter downstream at ingestion/processing time.

By the way, just declaring the scraped data to be CC0 may not necessarily be enough for every downstream client, certainly not for data scraped from the web. Some people will want to verify such claims at source, for every single feed, and come up with some policy for dealing with data that can’t be traced. (One policy that I’ve seen in the past: Allow bulk imports only for data under double-checked open data licenses, and use scraped data — where websites could have opted out via `robots.txt` — only for checking the quality of the clean data set. I don’t think ATP can define policies, but ATP can make the decisions easier by giving the needed input). Admittely, I’ve worked at some places with really paranoid lawyers, but this seems to be quite common with geographic data. Of course, my personal experience doesn’t really here...)

Another aspect that comes to mind: Machine-readable attribution, since some datasets require attribution and managing those is painful. (For example, there’s some open government data that gets released under Creative Commons). With tags like `attribution=City of Minnetonkaa`, `attribution:website=https://example.org/landing-page` etc., downstream clients could auto-generate attribution pages like [this](https://www.openstreetmap.org/copyright#contributors) or [that](https://support.google.com/waze/answer/12075833). It’s not exactly the biggest problem on the planet, but some hypothetical clients might be helpful.

**Comment by mjoe999** [comment]
Your attribution comment is well made. And indeed, yes, some people / lawyers get rather paranoid. For lawyers ... it's their job ... more worry, more time, more money :-) And as you say, at the end of the day, it's the responsibility of any downstream users to satisfy themselves with their use of the data. In the real-world most data owners are not looking to "hide" but wish to be represented well in downstream applications, where the more vanilla the downstream application is the better i.e. inviting user comments on a POI, price comparison etc start to stray into the I'm upset with you and I may do something about it territory. Giving something back to brands/websites is where we should look to be (the chat on https://github.com/alltheplaces/alltheplaces/issues/4310 is relevant here).
It's a shame how sloppy some of the sites are. I don't think it is the intention of the management!

**Comment by Cj-Malone** [comment]
I think at one point we'd have said that brand/Wikidata are spider level tags. And for most of them, they are. But we have many spiders now that have multiple brands and QIDs in there output. I'm not against it being in the GeoJson header/footer, but I think we should document that if those keys exist in the feature, they override the feed values.

**Comment by Cj-Malone** [comment]
Implemented in #4339
